sw/device/tests/rstmgr_sw_rst_ctrl_test.c

```diff
@@ -0,0 +1,308 @@
+ * value') when sw resets are asserted, those values are updated to reset value.
+ * In each round, the rstmgr asserts reset to each peripheral device and
+ * each test register value is compared with the expected reset value.
+ */
+
+typedef struct node node_t;
+typedef void config_f(node_t *);
+
+// peripheral device template
+// name : device name
+// base : base address
+// offset : offset to the test register
+// handler : dif handler for each peripheral device
+// config_f : function ptr to each peripheral program
+// reset_val : reset value for comparison
+// prgm_val : expected value for test register after configuration
+//            function is called.
+struct node {
+  const char *name;
+  uintptr_t base;
+  ptrdiff_t offset;
+  void *handler;
+  config_f *config;
+  uint32_t reset_val;
+  uint32_t prgm_val;
+};
+
+// handler declaration
+static dif_rstmgr_t rstmgr;
+static dif_spi_device_t spi_dev;
+static dif_spi_host_t spi_host0, spi_host1;
+static dif_usbdev_t usbdev;
+static dif_i2c_t i2c0, i2c1, i2c2;
+
+#define kNumNodes kTopEarlgreyResetManagerSwResetsLast + 1
+// pre declaration of template
+static node_t test_node[kNumNodes];
+
+// test register read from a given peripheral device
+static uint32_t read_test_reg(node_t *node) {
+  return (mmio_region_read32(mmio_region_from_addr(node->base), node->offset));
+}
+
+// init function
+static void init_nodes(void) {
+#define INITIALIZE(module_, index_)                                          \
+  CHECK_DIF_OK(                                                              \
+      dif_##module_##_init(mmio_region_from_addr(test_node[index_].base),    \
+                           (dif_##module_##_t *)test_node[index_].handler)); \
+  test_node[index_].reset_val = read_test_reg(&test_node[index_]);           \
+  LOG_INFO("%s reset value=0x%x", test_node[index_].name,                    \
+           test_node[index_].reset_val)
+
+  INITIALIZE(spi_device, kTopEarlgreyResetManagerSwResetsSpiDevice);
+  INITIALIZE(spi_host, kTopEarlgreyResetManagerSwResetsSpiHost0);
+  INITIALIZE(spi_host, kTopEarlgreyResetManagerSwResetsSpiHost1);
+  INITIALIZE(usbdev, kTopEarlgreyResetManagerSwResetsUsb);
+  // USBIF INIT is covered by USB init
+  INITIALIZE(i2c, kTopEarlgreyResetManagerSwResetsI2c0);
+  INITIALIZE(i2c, kTopEarlgreyResetManagerSwResetsI2c1);
+  INITIALIZE(i2c, kTopEarlgreyResetManagerSwResetsI2c2);
+#undef INITIALIZE
+}
+
+// configuration function wrapper
+static void spi_dev_config(node_t *node) {
+  const uint16_t kFifoLen = 0x800;
+  const dif_spi_device_config_t cfg = {
+      .clock_polarity = kDifSpiDeviceEdgePositive,
+      .data_phase = kDifSpiDeviceEdgeNegative,
+      .tx_order = kDifSpiDeviceBitOrderLsbToMsb,  // 1
+      .rx_order = kDifSpiDeviceBitOrderLsbToMsb,  // 1
+      .rx_fifo_timeout = 63,
+      .rx_fifo_len = kFifoLen,
+      .tx_fifo_len = kFifoLen,
+  };
+  CHECK_DIF_OK(dif_spi_device_configure(&spi_dev, &cfg));
```

Comment:
You are passing `node` as argument to the function, which already registers the peripheral context, with `node->handler`. IMO it will be better to pass that instead of passing `&spi_dev`. Likewise in below functions.

Reply:
updated